sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -477,7 +477,10 @@ case class Add(
   override protected def withNewChildrenInternal(newLeft: Expression, newRight: Expression): Add =
     copy(left = newLeft, right = newRight)
 
-  override lazy val canonicalized: Expression = {
+  override lazy val canonicalized: Expression = dataType match {
```

Comment:
let's call `left.dataType`. `dataType` is a `def` and contains some calculation logic.

Reply:
My concern is that if left contains integer `Add`s and right contains decimal `Add`s, the result may still be different.

Reply:
how is that possible? The left and right must be same/compatible type.